File: lib/database/dbcarrotboard.py
# import the PostgreSQL adapter for Python
import psycopg2
from ruamel.yaml import YAML

yaml = YAML()

class DBcarrotboard:

    def __init__(self):
        
        # Loads the credential for the db
        data = self.load_db_login()
        dbname = data['dbname']
        user = data['user']
        password = data['password']
        host = data['host']
        port = data['port']
        
        self.table_name = 'CARROT_BOARD'

        # Connect to the PostgreSQL database server
        self.postgresConnection = psycopg2.connect(dbname = dbname, user=user, password=password, host=host, port=port)

        # Try to create a table if it doesn't exist

        try:
            if self.check_table(self.table_name) is False:
                self.create_table()
        except(Exception, psycopg2.DatabaseError) as error:
            print(error)
        finally:
            self.postgresConnection.commit()

    # Creates a table to store the data for the carrotBoard
    def create_table(self):
        # Get cursor object from the database connection
        cursor = self.postgresConnection.cursor()

        # Create table statement
        sqlCreateTable = '''CREATE TABLE CARROT_BOARD(
            CARROT_ID SERIAL PRIMARY KEY,
            EMOJI CHAR(20) NOT NULL,
            MESSAGE_ID INT NOT NULL,
            USER_ID INT NOT NULL,
            CHANNEL_ID INT NOT NULL,
            COUNT BIGINT,
            MESSAGE_CONTENTS CHAR(50)
            )'''

        try:
            # Create a table in PostgreSQL database
            cursor.execute(sqlCreateTable)
            cursor.close()
            self.postgresConnection.commit()
        
        except(Exception, psycopg2.DatabaseError) as error:
            print(error)
        
        finally:
            cursor.close()
            self.postgresConnection.commit()

    # Checks if the given table exists
    def check_table(self,table_name):
        cur = self.postgresConnection.cursor()
        cur.execute("select * from information_schema.tables where table_name=%s", (table_name,))
        return bool(cur.rowcount)

    # Count the number of entries in the db for a specific message
    def count_values(self,emoji,message_id, user_id, channel_id):
        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT count(*) from carrot_board where emoji = %s and message_id = %s and user_id = %s and channel_id = %s'''
        
        record_to_insert = (emoji, message_id,user_id, channel_id)
        cursor.execute(postgres_insert_query, record_to_insert)
        record = cursor.fetchall()
        cursor.close()
        return record[0][0]
    
    # Get the react count on a given message
    def get_count(self,emoji,message_id, user_id, channel_id):
        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT * from carrot_board where emoji = %s and message_id = %s and user_id = %s and channel_id = %s'''
        record_to_insert = (emoji, message_id, user_id, channel_id)

        cursor.execute(postgres_insert_query, record_to_insert)

        record = cursor.fetchone()
        cursor.close()

        # None check
        if record is None:
            return None
        
        return record[5]
 
    # Loads the db credentials from the file
    def load_db_login(self):
        with open('./data/config/database.yml') as file:
            settings = yaml.load(file)
        return settings

    # Inserts a given message to the database
    def add_value(self,emoji,message_id, user_id, channel_id,message_contents):

        count_val = self.count_values(emoji,message_id, user_id, channel_id)
        # Increase the count if the value exists else create a new value
        if count_val == 0 or count_val == None:
            cursor = self.postgresConnection.cursor()
            postgres_insert_query = ''' INSERT INTO carrot_board (EMOJI, MESSAGE_ID, USER_ID, CHANNEL_ID, COUNT, MESSAGE_CONTENTS) VALUES (%s,%s,%s,%s,%s,%s)'''
            record_to_insert = (emoji, message_id, user_id, channel_id,1, message_contents)
            cursor.execute(postgres_insert_query, record_to_insert)
            self.postgresConnection.commit()
            cursor.close()

        else:
            count = self.get_count(emoji,message_id, user_id, channel_id)
            count = count + 1
            cursor = self.postgresConnection.cursor()
            postgres_insert_query = ''' UPDATE carrot_board SET count = %s  where emoji = %s and message_id = %s and user_id = %s and channel_id = %s'''
            record_to_insert = (count,emoji, message_id, user_id, channel_id)
            cursor.execute(postgres_insert_query, record_to_insert)
            self.postgresConnection.commit()
            cursor.close()

    # Retrieve an entry by carrotboard_id
    def get_by_cb_id(self,cb_id):
        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT * from carrot_board where carrot_id = %s'''
        record_to_insert = (str(cb_id))
        cursor.execute(postgres_insert_query, record_to_insert)
        record = cursor.fetchone()
        cursor.close()
        return {
            'carrot_id':record[0],
            'emoji':record[1],
            'message_id':record[2],
            'user_id':record[3],
            'channel_id':record[4],
            'count':record[5],
            'contents': record[6].rstrip(" "),
        }

    # Retrieve an entry by message id
    def get_by_msg_emoji(self,message_id, emoji):
        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT * from carrot_board where message_id = %s and emoji = %s'''
        record_to_insert = (message_id, emoji)
        cursor.execute(postgres_insert_query, record_to_insert)
        record = cursor.fetchone()
        cursor.close()
        return {
            'carrot_id':record[0],
            'emoji':record[1],
            'message_id':record[2],
            'user_id':record[3],
            'channel_id':record[4],
            'count':record[5],
            'contents': record[6].rstrip(" "),
        }

    # Get all the messages of a given react above a minimum threshold
    def get_all(self,emoji, count_min):
        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT * from carrot_board where emoji = %s and count >= %s'''
        record_to_insert = (emoji, count_min)
        cursor.execute(postgres_insert_query, record_to_insert)
        records = cursor.fetchall()
        cursor.close()
        return records

    # Deletes an entry from the db - An entry is all the rows with the same message_id
    def del_entry(self, message_id, channel_id):
        cursor = self.postgresConnection.cursor()
        postgres_delete_query = '''DELETE FROM carrot_board where message_id = %s and channel_id = %s'''
        record_to_delete = (message_id, channel_id)
        cursor.execute(postgres_delete_query, record_to_delete)
        self.postgresConnection.commit()
        cursor.close()

    # Deletes the specific emoji entry for a given message
    def del_entry_emoji(self, emoji, message_id, user_id, channel_id):
        cursor = self.postgresConnection.cursor()
        postgres_delete_query = '''DELETE FROM carrot_board where message_id = %s and channel_id = %s  and user_id = %s and emoji = %s'''
        record_to_delete = (message_id, channel_id, user_id, emoji)
        cursor.execute(postgres_delete_query, record_to_delete)
        self.postgresConnection.commit()
        cursor.close()

    # Subtracts the count of an emoji for a given message id by 1
    def sub_value(self, emoji, message_id, user_id, channel_id):

            count = self.get_count(emoji, message_id, user_id, channel_id)
            
            if count is None:
                return
            
            elif (count - 1) <= 0:
                # remove from database
                self.del_entry_emoji(emoji,message_id,user_id, channel_id)
            else:
                count = count - 1
                cursor = self.postgresConnection.cursor()
                postgres_insert_query = ''' UPDATE carrot_board SET count = %s  where emoji = %s and message_id = %s and user_id = %s and channel_id = %s'''
                record_to_insert = (count, emoji, message_id, user_id, channel_id)
                cursor.execute(postgres_insert_query, record_to_insert)
                self.postgresConnection.commit()
                cursor.close()

    # Get all the messages by emoji
    def get_all_by_emoji(self, emoji, count_min):
        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT * from carrot_board where emoji = %s and count >= %s ORDER BY count DESC'''
        record_to_insert = (emoji, count_min)
        cursor.execute(postgres_insert_query, record_to_insert)
        records = cursor.fetchall()
        cursor.close()

        results = []
        for record in records:
            results.append(
                {
                    'carrot_id': record[0],
                    'emoji': record[1],
                    'message_id': record[2],
                    'user_id': record[3],
                    'channel_id': record[4],
                    'count': record[5],
                    'contents': record[6].rstrip(" "),
                }
            )

        # results.sort(key=lambda x: x["count"], reverse=True)
        return results
    
    # Get all the messages by user_id
    def get_all_by_user(self, emoji, count_min, user):

        cursor = self.postgresConnection.cursor()
        postgres_insert_query = ''' SELECT * from carrot_board where emoji = %s and count >= %s and user_id = %s ORDER BY count DESC'''
        record_to_insert = (emoji, count_min, user)
        cursor.execute(postgres_insert_query, record_to_insert)
        records = cursor.fetchall()
        cursor.close()

        results = []
        for record in records:
            results.append(
                {
                    'carrot_id': record[0],
                    'emoji': record[1],
                    'message_id': record[2],
                    'user_id': record[3],
                    'channel_id': record[4],
                    'count': record[5],
                    'contents': record[6].rstrip(" "),
                }
            )

        #results.sort(key=lambda x: x["count"], reverse=True)
        return results

File: lib/database/dbcarrotboard.js
const { Pool } = require("pg");
const yaml = require("js-yaml");
const fs = require("fs");
// const { count, table } = require('console');

// Class for the carrotboard db
class DBcarrotboard {
    constructor() {
        // Loads the db configuration
        const details = this.load_db_login();

        this.pool = new Pool({
            user: details["user"],
            password: details["password"],
            host: details["host"],
            port: details["port"],
            database: details["dbname"],
        });
        // The name of the table
        const table_name = "CARROT_BOARD";

        // Creates the table if it doesn't exists
        (async () => {
            const is_check = await this.check_table(table_name);
            // console.log(is_check);
            if (is_check == false) {
                await this.create_table();
            }
        })();
    }

    load_db_login() {
        // Get document, or throw exception on error
        try {
            const doc = yaml.load(fs.readFileSync("./config/database.yml"));
            return doc;
        } catch (e) {
            console.log(e);
        }
    }

    // Checks if the table exists in the db
    async check_table(table_name) {
        const client = await this.pool.connect();
        try {
            // console.log("Running check_table command")
            await client.query("BEGIN");
            const values = [table_name];
            const result = await client.query(
                "select * from information_schema.tables where table_name=$1",
                values,
            );
            await client.query("COMMIT");

            if (result.rowCount == 0) {
                return false;
            } else {
                return true;
            }
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // // console.log("Client released successfully.")
        }
    }

    // Creates a new table
    async create_table() {
        const client = await this.pool.connect();
        try {
            console.log("Running create_table_carrot_board");
            await client.query("BEGIN");
            const query = `CREATE TABLE CARROT_BOARD(
            CARROT_ID SERIAL PRIMARY KEY,
            EMOJI CHAR(40) NOT NULL,
            MESSAGE_ID BIGINT NOT NULL,
            USER_ID BIGINT NOT NULL,
            CHANNEL_ID BIGINT NOT NULL,
            COUNT BIGINT,
            MESSAGE_CONTENTS CHAR(50)
            )`;

            const result = await client.query(query);
            await client.query("COMMIT");
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.")
        }
    }

    async count_values(emoji, message_id, user_id, channel_id) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            //
            await client.query("BEGIN");

            const query =
                "SELECT count(*) from carrot_board where emoji = $1 and message_id = $2 and user_id = $3 and channel_id = $4";

            const values = [emoji, message_id, user_id, channel_id];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            return result["rows"][0]["count"];
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.")
        }
    }

    async get_count(emoji, message_id, user_id, channel_id) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            //
            await client.query("BEGIN");

            const query =
                "SELECT * from carrot_board where emoji = $1 and message_id = $2 and user_id = $3 and channel_id = $4";

            const values = [emoji, message_id, user_id, channel_id];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            return result.rows[0].count;
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.")
        }
    }

    async add_value(emoji, message_id, user_id, channel_id, message_contents) {
        const client = await this.pool.connect();
        try {
            await client.query("BEGIN");
            const count_val = await this.count_values(emoji, message_id, user_id, channel_id);
            // console.log(count_val)

            if (count_val == 0) {
                const query =
                    "INSERT INTO carrot_board (EMOJI, MESSAGE_ID, USER_ID, CHANNEL_ID, COUNT, MESSAGE_CONTENTS) VALUES ($1,$2,$3,$4,$5,$6)";
                const values = [emoji, message_id, user_id, channel_id, 1, message_contents];

                const result = await client.query(query, values);
                await client.query("COMMIT");
            } else {
                let count = await this.get_count(emoji, message_id, user_id, channel_id);
                count = Number(count) + 1;
                const query =
                    "UPDATE carrot_board SET count = $1  where emoji = $2 and message_id = $3 and user_id = $4 and channel_id = $5";
                const values = [count, emoji, message_id, user_id, channel_id];

                const result = await client.query(query, values);
                await client.query("COMMIT");
            }
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.")
        }
    }

    async get_by_cb_id(cb_id) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            //
            await client.query("BEGIN");

            const query = "SELECT * from carrot_board where carrot_id = $1";

            const values = [cb_id];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            if (result.rowCount == 0) {
                return null;
            }
            return {
                carrot_id: result.rows[0]["carrot_id"],
                emoji: result.rows[0]["emoji"].trim(),
                message_id: result.rows[0]["message_id"],
                user_id: result.rows[0]["user_id"],
                channel_id: result.rows[0]["channel_id"],
                count: result.rows[0]["count"],
                message_contents: result.rows[0]["message_contents"].trim(),
            };
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.")
        }
    }

    async get_by_msg_emoji(message_id, emoji) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            await client.query("BEGIN");

            const query = "SELECT * from carrot_board where message_id = $1 and emoji = $2";

            const values = [message_id, emoji];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            if (result.rowCount == 0) {
                return null;
            }
            return {
                carrot_id: result.rows[0]["carrot_id"],
                emoji: result.rows[0]["emoji"].trim(),
                message_id: result.rows[0]["message_id"],
                user_id: result.rows[0]["user_id"],
                channel_id: result.rows[0]["channel_id"],
                count: result.rows[0]["count"],
                message_contents: result.rows[0]["message_contents"].trim(),
            };
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }

    async get_all(count_min) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            await client.query("BEGIN");

            const query = "SELECT * from carrot_board where count >= $2";

            const values = [count_min];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            return result.rows;
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }

    async del_entry(message_id, channel_id) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            await client.query("BEGIN");

            const query = "DELETE FROM carrot_board where message_id = $1 and channel_id = $2";

            const values = [message_id, channel_id];
            const result = await client.query(query, values);
            await client.query("COMMIT");
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }

    async del_entry_emoji(emoji, message_id, user_id, channel_id) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            await client.query("BEGIN");

            const query =
                "DELETE FROM carrot_board where message_id = $1 and channel_id = $2  and user_id = $3 and emoji = $4";

            const values = [message_id, channel_id, user_id, emoji];
            const result = await client.query(query, values);
            await client.query("COMMIT");
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }

    async sub_value(emoji, message_id, user_id, channel_id) {
        const client = await this.pool.connect();
        try {
            let count = await this.get_count(emoji, message_id, user_id, channel_id);

            if (count - 1 <= 0) {
                this.del_entry_emoji(emoji, message_id, user_id, channel_id);
            } else {
                // console.log("Connected successfully.")
                await client.query("BEGIN");

                count = count - 1;
                const query =
                    "UPDATE carrot_board SET count = $1  where emoji = $2 and message_id = $3 and user_id = $4 and channel_id = $5";

                const values = [count, emoji, message_id, user_id, channel_id];
                const result = await client.query(query, values);
                await client.query("COMMIT");
            }
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }

    async get_all_by_emoji(emoji, count_min) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            await client.query("BEGIN");

            const query =
                "SELECT * from carrot_board where emoji = $1 and count >= $2 ORDER BY count DESC";

            const values = [emoji, count_min];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            return result.rows;
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }

    async get_all_by_user(emoji, count_min, user) {
        const client = await this.pool.connect();
        try {
            // console.log("Connected successfully.")
            await client.query("BEGIN");

            const query =
                "SELECT * from carrot_board where emoji = $1 and count >= $2 and user_id = $3 ORDER BY count DESC";

            const values = [emoji, count_min, user];
            const result = await client.query(query, values);
            await client.query("COMMIT");

            return result.rows;
        } catch (ex) {
            console.log(`Something wrong happend ${ex}`);
        } finally {
            await client.query("ROLLBACK");
            client.release();
            // console.log("Client released successfully.");
        }
    }
}

module.exports = {
    DBcarrotboard,
};

// Anonymous function for testing purposes
/*
(async () => {
    let db = new DBcarrotboard();
    //console.log(user)
})();

*/
